tidy3d/components/data/sim_data.py

```diff
@@ -978,9 +982,158 @@ def make_adjoint_sim(
+    ) -> tuple[list[Source], xr.DataArray]:
+        """Process adjoint sources for the case of one adjoint source at several freqs."""
+
+        # map of monitor name to set of unique frequencies in the corresponding data
+        sources_unique_freqs = {
+            key: {src.source_time.freq0 for src in val} for key, val in sources_adj_dict.items()
+        }
+
+        def error_msg_pre(mnt_name: str) -> str:
+            """First part of error message for specific monitor."""
+            return f"Can't compute adjoint source for data from monitor '{mnt_name}'. "
+
+        error_msg_post = (
+            "Can only compute adjoint source for several monitor output if each"
+            "data has one frequency only."
+        )
+
+        # perform validation of the adjoint sources
+
+        # first, make sure each monitor data only has one frequency
+        for mnt_name, freqs in sources_unique_freqs.items():
+            if len(freqs) > 1:
+                raise ValueError(
+                    error_msg_pre(mnt_name)
+                    + f"Monitor has {len(freqs)} frequencies. "
+                    + error_msg_post
+                )
+
+        # then, more generally, ensure that all monitor data have the same frequency
+        all_freqs = set()
+        for mnt_name, freqs in sources_unique_freqs.items():
+            freq = tuple(freqs)[0]
+            all_freqs.add(freq)
+            if len(all_freqs) > 1:
+                raise ValueError(
+                    error_msg_pre(mnt_name)
+                    + error_msg_post
+                    + " Detected different frequencies in different monitors."
+                )
+
+        log.info("Several adjoint sources from different monitors, all with same single frequency.")
+
+        # passed validation, put all sources into a single list and set None for post-normalize amps
+        adj_srcs = []
+        for srcs in sources_adj_dict.values():
+            adj_srcs += srcs
+
+        return adj_srcs, None
+
+    def _process_adjoint_sources_broadband(
+        self, sources_adj_dict: dict[str, list[Source]]
+    ) -> tuple[list[Source], xr.DataArray]:
+        """Process adjoint sources for the case of several sources at the same freq."""
+
+        adj_srcs = list(sources_adj_dict.values())[0]
+
+        src_broadband = self._make_broadband_source(adj_srcs=adj_srcs)
+        post_norm_amps = self._make_post_norm_amps(adj_srcs=adj_srcs)
 
-    def make_adjoint_sources(self, data_vjp_paths: set[tuple]) -> list[Source]:
+        log.info(
+            "Several adjoint sources, from one monitor. "
+            "Only difference between them is the source time. "
+            "Constructing broadband adjoint source and performing post-run normalization "
+            f"of fields with {len(post_norm_amps)} frequencies."
+        )
+
+        return [src_broadband], post_norm_amps
+
+    def _make_broadband_source(self, adj_srcs: List[Source], num_fwidth: float = 0.5) -> Source:
```

Comment:
Do we use `list` or `typing.List`? Collections like `list` can be used as generic types since Python 3.9, so I guess we should probably switch to using that everywhere since we don't support <3.9? I'll also have to change all of my type hints...

Reply:
yea good point, I'll change